server/datastore/mysql/software.go

```diff
@@ -2877,301 +3144,601 @@ last_vpp_install AS (
+			softwareTitle := bySoftwareTitleID[softwareTitleRecord.ID]
+
+			if softwareTitleRecord.SoftwareIDList != nil {
+				softwareIDList := strings.Split(*softwareTitleRecord.SoftwareIDList, ",")
+				softwareSourceList := strings.Split(*softwareTitleRecord.SoftwareSourceList, ",")
+				softwareVersionList := strings.Split(*softwareTitleRecord.VersionList, ",")
+				softwareBundleIdentifierList := strings.Split(*softwareTitleRecord.BundleIdentifierList, ",")
+
+				for index, softwareIdStr := range softwareIDList {
+					version := &fleet.HostSoftwareInstalledVersion{}
+
+					if softwareId, err := strconv.ParseUint(softwareIdStr, 10, 32); err == nil {
+
+						softwareId := uint(softwareId)
+						if software, ok := bySoftwareID[softwareId]; ok {
+							version.Version = softwareVersionList[index]
+							version.BundleIdentifier = softwareBundleIdentifierList[index]
+							version.Source = softwareSourceList[index]
+							version.LastOpenedAt = software.LastOpenedAt
+							version.SoftwareID = softwareId
+							version.SoftwareTitleID = softwareTitleRecord.ID
+
+							version.InstalledPaths = installedPathBySoftwareId[softwareId]
+							version.Vulnerabilities = vulnerabilitiesBySoftwareID[softwareId]
 
-			// store the CVEs with the proper software entry
-			for _, cve := range softwareCVEs {
-				ver := bySoftwareID[cve.SoftwareID]
-				ver.Vulnerabilities = append(ver.Vulnerabilities, cve.CVE)
+							if version.Source == "apps" {
+								version.SignatureInformation = pathSignatureInformation[softwareId]
+							}
+
+							if storedIndex, ok := indexOfSoftwareTitle[softwareTitleRecord.ID]; ok {
+								deduplicatedList[storedIndex].InstalledVersions = append(deduplicatedList[storedIndex].InstalledVersions, version)
+							} else {
+								softwareTitleRecord.InstalledVersions = append(softwareTitleRecord.InstalledVersions, version)
+							}
+						}
+					}
+				}
 			}
 
-			const pathsStmt = `
-			SELECT
-				hsip.software_id,
-				hsip.installed_path,
-				hsip.team_identifier
-			FROM
-				host_software_installed_paths hsip
-			WHERE
-				hsip.host_id = ? AND
-				hsip.software_id IN (?)
-			ORDER BY
-				software_id, installed_path
-	`
-			type installedPath struct {
-				SoftwareID     uint   `db:"software_id"`
-				InstalledPath  string `db:"installed_path"`
-				TeamIdentifier string `db:"team_identifier"`
+			if softwareTitleRecord.VPPAppAdamIDList != nil {
+				vppAppAdamIDList := strings.Split(*softwareTitleRecord.VPPAppAdamIDList, ",")
+				vppAppSelfServiceList := strings.Split(*softwareTitleRecord.VPPAppSelfServiceList, ",")
+				vppAppVersionList := strings.Split(*softwareTitleRecord.VPPAppVersionList, ",")
+				vppAppPlatformList := strings.Split(*softwareTitleRecord.VPPAppPlatformList, ",")
+				vppAppIconURLList := strings.Split(*softwareTitleRecord.VPPAppIconUrlList, ",")
+
+				if len(vppAppAdamIDList) > 1 {
+					fmt.Print("What are we going to do here???")
```

Comment:
👀

Reply:
A software title will only ever have one VPP app associated with it, and to my knowledge you can only install a VPP app once per system (/cc @jahzielv in case I'm wrong here). Not sure why we have a list here rather than a single value, but we can operate assuming a list length 1 if we must have a list.

Reply:
@iansltx you can issue multiple installs, but it will only install a single app on the host if that makes sense.